storage/src/lib.rs

```diff
@@ -0,0 +1,5006 @@
+    /// if there is a live version of the object.
+    pub fn if_generation_not_match(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("if_generation_not_match".into(), vec![value.into()]);
+        self
+    }
+    /// Makes the operation conditional on whether the object's current
+    /// metageneration matches the given value.
+    pub fn if_metageneration_match(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("if_metageneration_match".into(), vec![value.into()]);
+        self
+    }
+    /// Makes the operation conditional on whether the object's current
+    /// metageneration does not match the given value.
+    pub fn if_metageneration_not_match(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("if_metageneration_not_match".into(), vec![value.into()]);
+        self
+    }
+    /// Set of properties to return. Defaults to noAcl.
+    pub fn projection(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("projection".into(), vec![value.into()]);
+        self
+    }
+    /// The project to be billed for this request if the target bucket is
+    /// requester-pays bucket.
+    pub fn provisional_user_project(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("provisional_user_project".into(), vec![value.into()]);
+        self
+    }
+    /// The project to be billed for this request. Required for Requester
+    /// Pays buckets.
+    pub fn user_project(mut self, value: impl Into<String>) -> Self {
+        self.url_params
+            .insert("user_project".into(), vec![value.into()]);
+        self
+    }
+
+    pub async fn execute(self) -> Result<model::Object> {
+        let client = self.client.inner;
+        let tok = client.cred.access_token().await.map_err(Error::wrap)?;
+        let res = client
+            .http_client
+            .get(format!(
+                "{}b/{}/o/{}",
+                client.base_path,
+                self.bucket.unwrap(),
+                self.object.unwrap()
+            ))
+            .query(
+                &self
+                    .url_params
+                    .iter()
+                    .map(|(k, v)| (k.as_str(), v[0].as_str()))
+                    .collect::<Vec<(&str, &str)>>(),
+            )
+            .query(&[("alt", "json")])
+            .query(&[("prettyPrint", "false")])
+            .bearer_auth(tok.value)
+            .send()
+            .await
+            .map_err(Error::wrap)?;
+        if !res.status().is_success() {
+            let error: ApiErrorReply = res.json().await.map_err(Error::wrap)?;
+            return Err(Error::wrap(error.into_inner()));
+        }
+        let res: model::Object = res.json().await.map_err(Error::wrap)?;
+        Ok(res)
+    }
+
+    pub async fn download(self) -> Result<Vec<u8>> {
```

Comment:
@SurferJeffAtGoogle We need a better return type here, and I don't know if AsyncRead is good enough. I feel like we should return a concrete type folks can work with easier. Maybe this is where something like the BytesReader type I made make sense? Thoughts?

Reply:
Maybe a [BufReader](https://docs.rs/tokio/latest/tokio/io/struct.BufReader.html)?

Reply:
What would you suggest for a return value here?

Reply:
Return a `Box<AsyncRead>`.

Reply:
I am a little concerned about the usability of that type I guess. Unless the user also pulls in [AsyncReadExt](https://docs.rs/tokio/1.15.0/tokio/io/trait.AsyncReadExt.html) that type is quite hard to use imo. I will move to this for now but I think this needs more disscussion.

Reply:
> I am a little concerned about the usability of that type I guess. Unless the user also pulls in [AsyncReadExt](https://docs.rs/tokio/1.15.0/tokio/io/trait.AsyncReadExt.html) that type is quite hard to use imo. I will move to this for now but I think this needs more disscussion.

I was 100% assuming that the caller would import `AsyncReadExt`.  I couldn't imagine using `AsyncRead` directly.

Reply:
I think I found a better API for this, and that is accepting an AsyncWrite instead. I will make this change and update the examples so you can LMK what you think

Reply:
Ok.  I'll accept this code if you open a bug saying that download() must not store the entire payload in memory.

Reply:
Ack, opened #7